pouta_blueprints/rules.py

```diff
@@ -0,0 +1,110 @@
+from pouta_blueprints.models import Blueprint, BlueprintTemplate, Instance, GroupUserAssociation
+from pouta_blueprints.views.commons import is_group_manager
+from sqlalchemy import or_, and_
+from sqlalchemy.orm import load_only
+from sqlalchemy.sql.expression import true
+import itertools
+# import logging
+
+
+def apply_rules_blueprint_templates(user, args=None):
+
+    q = BlueprintTemplate.query
+    if not user.is_admin:
+        query_exp = BlueprintTemplate.is_enabled == true()
+        q = q.filter(query_exp)
+    if args is not None and 'template_id' in args:
+        q = q.filter_by(id=args.get('template_id'))
+
+    return q
+
+
+def apply_rules_blueprints(user, args=None):
+
+    q = Blueprint.query
+    if not user.is_admin:
+        group_user_objs = GroupUserAssociation.query.filter_by(user_id=user.id, manager=False).all()
+        user_group_ids = [group_user_obj.group.id for group_user_obj in group_user_objs]
+        banned_group_ids = [banned_group_item.id for banned_group_item in user.banned_groups.all()]
+        allowed_group_ids = set(user_group_ids) - set(banned_group_ids)  # do not allow the banned users
+
+        # Start building query expressions based on the condition that :
+        # a group manager can see all of his blueprints and only enabled ones of other groups
+        query_exp = Blueprint.is_enabled == true()
+        allowed_group_ids_exp = None
+        if allowed_group_ids:
+            allowed_group_ids_exp = Blueprint.group_id.in_(allowed_group_ids)
+        query_exp = and_(allowed_group_ids_exp, query_exp)
+
+        manager_group_ids = get_manager_group_ids(user)
+        manager_group_ids_exp = None
+        if manager_group_ids:
+            manager_group_ids_exp = Blueprint.group_id.in_(manager_group_ids)
+        query_exp = or_(query_exp, manager_group_ids_exp)
+        q = q.filter(query_exp)
+
+    if args is not None and 'blueprint_id' in args:
+        q = q.filter_by(id=args.get('blueprint_id'))
+
+    return q
+
+
+def apply_rules_export_blueprints(user):
+    q = Blueprint.query
```

Comment:
Or a docstring as well. This one is almost quickly legible but even then reasoning about DB api code can be something of a cognitive load.